rust/rope/src/spans.rs

```diff
@@ -301,21 +301,30 @@ impl<T: Clone> Spans<T> {
         *self = b.build();
     }
 
-    // FIXME: Instead of iterating through all spans every time, another option would be to go
-    // leaf-by-leaf, and check each leaf for whether or not it has any items in the interval;
-    // if they don't we keep them unchanged, otherwise we do this operation, but only within the leaf.
-    //
     /// Deletes all spans that intersect with `interval`.
     pub fn delete_intersecting(&mut self, interval: Interval) {
-        let mut builder = SpansBuilder::new(self.len());
-        for (iv, data) in self.iter() {
-            // check if spans overlaps with interval
-            if iv.intersect(interval).is_empty() {
-                // keep the ones that are not overlapping
-                builder.add_span(iv, data.clone());
+        let mut b = TreeBuilder::new();
+        // the end of the last leaf seen
+        let mut end_offset = 0;
+        let mut leaf_iterator = self.iter_leaf_nodes();
+        while let Some((node, leaf)) = leaf_iterator.next() {
+            let SpansLeaf { len, spans } = leaf;
+            if spans.iter().any(|s| !s.iv.intersect(interval).is_empty()) {
+                let spans: Vec<_> =
+                    spans.iter().filter(|s| s.iv.intersect(interval).is_empty()).cloned().collect();
+                b.push_leaf(SpansLeaf { spans, len: *len });
+            } else {
+                b.push(node.clone());
+            }
+
+            // stop looking at leaves if we're past the end of the deletion interval
+            end_offset += node.len();
+            if end_offset > interval.end() {
+                break;
             }
         }
-        *self = builder.build();
+        b.push(self.subseq(interval.end..));
```

Comment:
In testing the subseq function doesn't seem to work as intended.
In the `delete_intersecting_empty` test `self.subseq(interval.end..)` returns a leaf with one span, iv=(2,3)